### PR TITLE
Fix crash in dns.lookup, ensure getaddrinfo() only returns IPv4-only or IPv6-only results when it should, normalize node:dns errors

### LINKED ISSUE #6658: fetch fails on https://litudian.com ( works on curl/fetch)
### What version of Bun is running?

1.0.7

### What platform is your computer?

Darwin 22.6.0 arm64 arm

### What steps can reproduce the bug?

const res = await fetch(“https://litudian.com”, {
    keepalive:false,
});

console.log(res);

### What is the expected behavior?

it should return a 200 

### What do you see instead?

ConnectionRefused: Unable to connect. Is the computer able to access the url? 

### Additional information

we also have this issue on bun docker image

### LINKED ISSUE #7298: `bun create elysia myapp` takes a long time and then returns `connection refused`
### What version of Bun is running?

1.0.14+d8be3e51b

### What platform is your computer?

Darwin 21.6.0 x86_64 i386

### What steps can reproduce the bug?

I just type ```bun create elysia myapp```

### What is the expected behavior?

It should create a new app with Elysia framework

### What do you see instead?

```



error: ConnectionRefused

----- bun meta -----
Bun v1.0.14 (d8be3e51) macOS x64 21.6.0
CreateCommand: 
Elapsed: 75023ms | User: 26ms | Sys: 14ms
RSS: 7.18MB | Peak: 7.18MB | Commit: 34.61MB | Faults: 0
----- bun meta -----

0   0x106304860

Search GitHub issues https://bun.sh/issues or ask for #help in https://bun.sh/discord

```

### Additional information

 - I tried many times, there is not variation.
 - My internet connection is fine, here is a sample
 ```
renatoexpert@MacBook-Air ~ % ping google.com
PING google.com (142.251.128.142): 56 data bytes
64 bytes from 142.251.128.142: icmp_seq=0 ttl=58 time=85.978 ms
64 bytes from 142.251.128.142: icmp_seq=1 ttl=58 time=112.455 ms
64 bytes from 142.251.128.142: icmp_seq=2 ttl=58 time=66.558 ms
64 bytes from 142.251.128.142: icmp_seq=3 ttl=58 time=79.403 ms
64 bytes from 142.251.128.142: icmp_seq=4 ttl=58 time=74.697 ms
64 bytes from 142.251.128.142: icmp_seq=5 ttl=58 time=105.295 ms
64 bytes from 142.251.128.142: icmp_seq=6 ttl=58 time=41.033 ms
64 bytes from 142.251.128.142: icmp_seq=7 ttl=58 time=72.546 ms
64 bytes from 142.251.128.142: icmp_seq=8 ttl=58 time=116.214 ms
64 bytes from 142.251.128.142: icmp_seq=9 ttl=58 time=98.537 ms
64 bytes from 142.251.128.142: icmp_seq=10 ttl=58 time=122.181 ms
^C
--- google.com ping statistics ---
11 packets transmitted, 11 packets received, 0.0% packet loss
round-trip min/avg/max/stddev = 41.033/88.627/122.181/23.602 ms
```
 - Internet Provider is Starlink, and has no support to IPv4, only IPv6. This make some apps, like bank mobile apps to not connect or to take so much time.

### LINKED ISSUE #8789: Postgres.js connection error with bun >0.6.14
### What version of Bun is running?

1.0.26

### What platform is your computer?

Darwin 23.2.0 arm64

### What steps can reproduce the bug?

1. bun init
2. bun add postgres
3. configure a connection
4. try to connect and select data

### What is the expected behavior?

return data

### What do you see instead?

```
165 |   if (originCache.has(xs))
166 |     return originCache.get(xs)
167 |
168 |   const x = Error.stackTraceLimit
169 |   Error.stackTraceLimit = 4
170 |   originCache.set(xs, new Error())
                               ^
ECONNREFUSED: Failed to connect
 syscall: "connect"

      at cachedError (/Users/akha/WebstormProjects/BunPostgresError/node_modules/postgres/src/query.js:170:26)
      at new Query (/Users/akha/WebstormProjects/BunPostgresError/node_modules/postgres/src/query.js:36:24)
      at sql (/Users/akha/WebstormProjects/BunPostgresError/node_modules/postgres/src/index.js:112:15)
      at readLogs (/Users/akha/WebstormProjects/BunPostgresError/index.ts:31:31)
```

### Additional information

It works using bun 0.6.x
Doesn't work using bun 0.7.x, 0.8.x, 1.0.x.

### LINKED ISSUE #11205: ECONNREFUSED: Failed to connect when using `node-postgres` and `postgres` packages
### What version of Bun is running?

1.1.8+89d25807f

### What platform is your computer?

Darwin 23.4.0 arm64 arm

### What steps can reproduce the bug?

Create two files that attempt to connect to a pg db:

## When using the [`postgres` package](https://github.com/porsager/postgres?tab=readme-ov-file):

```import postgres from 'postgres';
const sql = postgres({
  // config vars here
});
const users = await sql`select * from users;`
```

## When using [`node-pg`](https://node-postgres.com/):

```import pg from "pg";

const { Pool } = pg;
const pool = new Pool({ max: 10 });

const makeDbClientQuery = (query, client) => {
  return client.query(query[0], query[1]);
};

const extractRows = (response) => response.rows;

export const exec = async (queries) => {
  const client = await pool.connect();

  try {
    await client.query("begin");
    const responses = await Promise.all(queries.map(query => makeDbClientQuery(query, client)));
    await client.query("commit");
    return responses.map(extractRows);
  } catch (e) {
    await client.query('rollback')
    throw e;
  } finally {
    client.release();
  }
}

const res = await exec([['select * from users;', '']]);
pool.end();
console.log(res)```

## Both libraries raise the following error:

```ECONNREFUSED: Failed to connect
 syscall: "connect"
 ```
 
> running the same code with node does not raise an error.

### What is the expected behavior?

Successful connection and query execution.

### What do you see instead?

```ECONNREFUSED: Failed to connect
 syscall: "connect"
 ```

### Additional information

_No response_

### LINKED ISSUE #11882: SegFault when using NATS on WSL2
### How can we reproduce the crash?

Nats v2.26.0
Bun v1.1.13

https://github.com/btmnk/test-bun-nats

It only segfaults on WSL2 with Ubuntu 20 and 22 (both LTS versions)

When I tried creating a repro on CodeSandbox it worked fine so I assume it has something to do with WSL.
I only got that error after upgrading to bun 1.1.13, it works fine when using 1.0.9.

### Relevant log output

```shell
============================================================
Bun v1.1.13 (bd6a6051) Linux x64
Args: "bun" "run" "index.ts"
Features: jsc transpiler_cache tsconfig
Builtins: "bun:main" "node:dns" "node:fs" "node:net" "node:path" "node:string_decoder" "node:tls" "node:util/types"
Elapsed: 48ms | User: 42ms | Sys: 10ms
RSS: 1.07GB | Peak: 77.41MB | Commit: 1.07GB | Faults: 0

panic(main thread): Segmentation fault at address 0x20
oh no: Bun has crashed. This indicates a bug in Bun, not your code.

To send a redacted crash report to Bun's team,
please file a GitHub issue using the link below:

 https://bun.report/1.1.13/lr1bd6a605AiggggF2nk6kE+xyQ25i/jEA2AgC

[1]    692 illegal hardware instruction  bun run index.ts
```

### Stack Trace (bun.report)

Bun v1.1.13 ([`bd6a605`](<https://github.com/oven-sh/bun/tree/bd6a6051203ff577240615ff4e242270035a36b7>)) on linux x86_64 [RunCommand]

Segmentation fault at address 0x00000020

- [`ld-temp.o:0`](<https://github.com/oven-sh/bun/blob/bd6a6051203ff577240615ff4e242270035a36b7/ld-temp.o#L0>): `WTF::jscSignalHandler`
- `??`
- [`ld-temp.o:0`](<https://github.com/oven-sh/bun/blob/bd6a6051203ff577240615ff4e242270035a36b7/ld-temp.o#L0>): `end_hquery`

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
### What does this PR do?

This does three things:
1) Fixes a crash in dns.lookup
2) On IPv6-only networks or IPv4-only networks, `getaddrinfo()` will now only return connectable addresses due to the 
3) Normalize DNS errors to better match Node. Node treats `ENONAME`, `ENODATA` as both `ENOTFOUND`.

- Fixes #11882
- Fixes #8789
- Fixes #11205
- Fixes #6658 (may already have been fixed by happy eyeballs, but hard to tell)
- Fixes #7298 (may already have been fixed by happy eyeballs, but hard to tell)

Duplicates:
- Fixes https://github.com/oven-sh/bun/issues/12021
- Fixes https://github.com/oven-sh/bun/issues/11890
- Fixes https://github.com/oven-sh/bun/issues/11810

Not as confident this fixes these, but there's a high chance it does:
- https://github.com/oven-sh/bun/issues/11423
- https://github.com/oven-sh/bun/issues/11223
- https://github.com/oven-sh/bun/issues/4066

### How did you verify your code works?

Re-enabled some tests, but that crash in CI should be fixed

### CODE REVIEW

**Comment by robobun** [comment]
@Jarred-Sumner, your commit 4ea4f4854f52f9db04067c45fc5e37da21be4ded is building: [`#432`](https://buildkite.com/bun/bun/builds/432)

File: test/js/bun/dns/resolve-dns.test.ts [hunk 1]
```diff
@@ -6,7 +6,7 @@ import { isIP, isIPv4, isIPv6 } from "node:net";
 const backends = ["system", "libc", "c-ares"];
 const validHostnames = ["localhost", "example.com"];
 const invalidHostnames = ["adsfa.asdfasdf.asdf.com"]; // known invalid
-const malformedHostnames = ["", " ", ".", " .", "localhost:80", "this is not a hostname"];
+const malformedHostnames = [" ", ".", " .", "localhost:80", "this is not a hostname"];
```
Comment (Electroid) [inline test/js/bun/dns/resolve-dns.test.ts:9]: Why is the empty case removed?

File: src/bun.js/api/bun/dns_resolver.zig [hunk 2]
```diff
@@ -271,6 +271,10 @@ pub fn normalizeDNSName(name: []const u8, backend: *GetAddrInfo.Backend) []const
         } else if (strings.isIPV6Address(name)) {
             backend.* = .system;
         }
+        // getaddrinfo() is inconsistent with ares_getaddrinfo() when using localhost
```
Comment (Electroid) [inline src/bun.js/api/bun/dns_resolver.zig:274]: This doesn't really explain anything, what exactly is inconsistent?

File: test/js/bun/dns/resolve-dns.test.ts (hunk 1 above)
Comment (Electroid) [reply test/js/bun/dns/resolve-dns.test.ts:9]: If this is because of the new case for `ENODATA`, I don't see a test for it.

Comment (Jarred-Sumner) [reply test/js/bun/dns/resolve-dns.test.ts:9]: It returns a different error on empty string, and previously `malformedHostnames` was not used at all

**Pull request merged**